wp-includes/class-wp-https-ui.php

```diff
@@ -0,0 +1,336 @@
+	 */
+	public function __construct( $wp_https_detection ) {
+		$this->wp_https_detection = $wp_https_detection;
+	}
+
+	/**
+	 * Initializes the object.
+	 */
+	public function init() {
+		add_action( 'admin_init', array( $this, 'init_admin' ) );
+		add_action( 'init', array( $this, 'filter_site_url_and_home' ) );
+		add_action( 'init', array( $this, 'filter_header' ) );
+	}
+
+	/**
+	 * Initializes the admin tasks.
+	 */
+	public function init_admin() {
+		$this->register_settings();
+		$this->add_settings_field();
+	}
+
+	/**
+	 * Registers the HTTPS settings.
+	 */
+	public function register_settings() {
+		register_setting(
+			self::OPTION_GROUP,
+			self::UPGRADE_HTTPS_OPTION,
+			array(
+				'type'              => 'string',
+				'sanitize_callback' => array( $this, 'upgrade_https_sanitize_callback' ),
+			)
+		);
+	}
+
+	/**
+	 * Sanitization callback for the upgrade HTTPS option.
+	 *
+	 * @param string $raw_value The value to sanitize.
+	 * @return bool Whether the option is true or false.
+	 */
+	public function upgrade_https_sanitize_callback( $raw_value ) {
+		unset( $raw_value );
+		return isset( $_POST[ self::UPGRADE_HTTPS_OPTION ] ); // WPCS: CSRF OK.
+	}
+
+	/**
+	 * Adds a settings field to the 'General Settings' page.
+	 *
+	 * Only add this if the site can support HTTPS,
+	 * but the home and siteurl option values are not HTTPS (WordPress Address and Site Address).
+	 * This UI would not apply if those URLs are already HTTPS.
+	 */
+	public function add_settings_field() {
+		if ( ! $this->wp_https_detection->is_currently_https() ) {
+			add_settings_field(
+				self::HTTPS_SETTING_ID,
+				__( 'HTTPS', 'pwa' ),
+				array( $this, 'render_https_settings' ),
+				self::OPTION_GROUP
+			);
+		}
+	}
+
+	/**
+	 * Renders the HTTPS settings in /wp-admin on the General Settings page.
+	 */
+	public function render_https_settings() {
+		$upgrade_https_value = (bool) get_option( self::UPGRADE_HTTPS_OPTION );
+		$https_more_details  = sprintf(
+			'<a href="%s">%s</a>',
+			__( 'https://make.wordpress.org/support/user-manual/web-publishing/https-for-wordpress/', 'pwa' ),
+			__( 'More details', 'pwa' )
+		);
+
+		?>
+		<p class="description">
+			<?php
+			echo wp_kses_post( sprintf(
+				/* translators: %s: a link for more details */
+				__( 'HTTPS is essential to securing your WordPress site, we strongly suggest upgrading to it. %s', 'pwa' ),
+				$https_more_details
+			) );
+			?>
+		</p>
+		<p>
+			<label><input name="<?php echo esc_attr( self::UPGRADE_HTTPS_OPTION ); ?>" type="checkbox" <?php checked( $upgrade_https_value ); ?> value="<?php echo esc_attr( self::OPTION_CHECKED_VALUE ); ?>"><?php esc_html_e( 'Upgrade to secure connection', 'pwa' ); ?></label>
+		</p>
+		<script>
+			( function ( $ ) {
+				// Move this UI under the Site Address (URL) <tr> on the General Settings page.
+				$( 'input[name=<?php echo esc_attr( self::UPGRADE_HTTPS_OPTION ); ?>]' ).parents( 'tr' ).insertAfter( $( 'label[for=home]' ).parents( 'tr') );
```

Comment:
Use of `esc_attr` isn't right here. It should be `wp_json_encode()`, kinda, but really it's not necessary.

Reply:
Sure, https://github.com/xwp/pwa-wp/pull/36/commits/9e969c0475549792f504528a6e1041b9440a2262 removes the escaping here.